src/ui/public/vislib/lib/axis_config.js

```diff
@@ -0,0 +1,108 @@
+import _ from 'lodash';
+export default function AxisConfigFactory() {
+
+  const defaults = {
+    show: true,
+    type: 'value',
+    elSelector: '.axis-wrapper-{pos} .axis-div',
+    position: 'left',
+    scale: {
+      type: 'linear',
+      expandLastBucket: true, //TODO: rename ... bucket has nothing to do with vis
+      inverted: false
+    },
+    style: {
+      color: '#ddd',
+      lineWidth: '1px',
+      opacity: 1,
+      tickColor: '#ddd',
+      tickWidth: '1px',
+      tickLength: '6px'
+    },
+    labels: {
+      axisFormatter: null, // TODO: create default axis formatter
+      show: true,
+      rotate: 0,
+      rotateAnchor: 'center',
+      filter: false,
+      color: '#ddd',
+      font: '"Open Sans", "Lato", "Helvetica Neue", Helvetica, Arial, sans-serif', // TODO
+      fontSize: '8pt',
+      truncate: 100
+    },
+    title: {
+      text: '',
+      elSelector: '.axis-wrapper-{pos} .axis-title'
+    }
+  };
+
+  const categoryDefaults = {
+    type: 'category',
+    position: 'bottom',
+    labels: {
+      rotate: 0,
+      rotateAnchor: 'end',
+      filter: true
+    }
+  };
+
+  class AxisConfig {
+    constructor(config) {
+      if (config.type === 'category') {
+        _.defaultsDeep(config, categoryDefaults);
+      }
+      _.defaultsDeep(config, defaults);
+      _.extend(this, config);
+
+      this.elSelector = this.elSelector.replace('{pos}', this.position);
+      this.rootEl = this.vis.el;
+
+      if (this.type === 'category') {
+        this.values = this.data.xValues();
+        this.ordered = this.data.get('ordered');
+      }
+    };
+
+    get(property) {
+      return _.get(this, property, null);
```

Comment:
how exactly would that work ? (has + defaults) ...
if we have a defaults when there isnt a value set we return default ...

or should we return error if value is not set at all, and only return default if value is set to undefined ?